### ISSUE TITLE
Inconsistent behaviour of rmarkdown::render() ouput_file= parameter while rendering

### ISSUE THREAD

**Issue by rifius** [rstudio/rmarkdown#1913]
Rendering an R Notebook document:

When specifying an output file to the `render()` function, the `.nb.html` is appended to the supplied file name inconsistently.
1. If the base name does not contain a dot or an extension, the file name is appended with `.nb.html`
2. If the base name contains an extension without an underscore (`_`) --and possibly other punctuation--, no `.nb.html` is appended.
3. If the extension does contain underscore (`_`), then `.nb.html` is appended.

I have not tested this with other output formats other than notebook. I think case 1 and case 3 are ok, but case 2 is inconsistent.
Always appending (or always not appending) would be a consistent behaviour, imho.  Perhaps checking if the `.nb.html` extension exists could be a reason to not append it.

MWE included:

```{r}
ll = "---\ntitle: \"Test\"\noutput: html_notebook\n---\n\n```{r}\nplot(cars)\n```\n"
f = tempfile(fileext=".Rmd")
writeLines(ll, con=f)
# Case 1: Appended
rmarkdown::render(input=f, output_file="A", output_dir=".", quiet=T)
# Case 2: Not appended
rmarkdown::render(input=f, output_file="A.b", output_dir=".", quiet=T)
# Case 3: Appended
rmarkdown::render(input=f, output_file="A.b_c", output_dir=".", quiet=T)
# Other tests
rmarkdown::render(input=f, output_file="A.verylongconformingextension", output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file="A.verylongnon-conformingextension", output_dir=".", quiet=T)
# These next two understandably fail
rmarkdown::render(input=f, output_file="Another.verylongnon$conformingextension", output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file="A.v$x", output_dir=".", quiet=T)
# But this one does not
rmarkdown::render(input=f, output_file="A.v=x", output_dir=".", quiet=T)

xfun::session_info('rmarkdown')
```

    R version 4.0.2 (2020-06-22)
    Platform: x86_64-redhat-linux-gnu (64-bit)
    Running under: Fedora 32 (Workstation Edition), RStudio 1.3.1093  
    Locale: 
      LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C               LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8
      LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8    LC_PAPER=en_AU.UTF-8       LC_NAME=C
      LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C         

    Package version:
      base64enc_0.1.3 digest_0.6.25   evaluate_0.14   glue_1.4.2      graphics_4.0.2  grDevices_4.0.2 highr_0.8       htmltools_0.5.0
      jsonlite_1.7.1  knitr_1.30      magrittr_1.5    markdown_1.1    methods_4.0.2   mime_0.9        rlang_0.4.7     rmarkdown_2.4.1   
      stats_4.0.2     stringi_1.5.3   stringr_1.4.0   tinytex_0.26    tools_4.0.2     utils_4.0.2     xfun_0.18       yaml_2.2.1   

    Pandoc version: 2.7.3

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by cderv** [comment]
Hi, 

Thank for the report. 

Per the doc, 

> If output_file is specified but does not have a file extension, an extension will be automatically added according to the output format.

So yes, if the value provided in `output_file` is not recognized as a filename with an extension, one will be added. 

How to the detection of extension works ? 

Currently, rmarkdown is using `xfun::file_ext` to determine extension of the output file, which is a wrapper of `tools::file_ext`. 
If not extension is detected, then an automatic one is added based on the output format, here `.nb.html` because you used `html_notebook`

From the doc of `tools::file_ext`, 
> file_ext returns the file (name) extensions (excluding the leading dot). **(Only purely alphanumeric extensions are recognized.)**

and this is what we observed
``` r
# Your example filenames
files <- c("A", "A.b", "A.b_c", "A.verylongconformingextension", 
           "A.verylongnon-conformingextension", 
           "Another.verylongnon$conformingextension", 
           "A.v$x", "A.v=x")
# use by rmarkdown
xfun::file_ext(files)
#> [1] ""                            "b"                          
#> [3] ""                            "verylongconformingextension"
#> [5] ""                            ""                           
#> [7] ""                            ""
# which is a wrapper of 
tools::file_ext(files)
#> [1] ""                            "b"                          
#> [3] ""                            "verylongconformingextension"
#> [5] ""                            ""                           
#> [7] ""                            ""
```

<sup>Created on 2020-10-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

So currently, the behavioir is consistent with R behavior regarding filename and extension. 

One question about this : 
Does other extensions than purely alphanumeric one exists ? I don't think I ever encounter an extension with a $ or a _ , and all the more in the Rmarkdown context where you will output html, pdf, docx, md.  I don't see why you would need to precise another extension with some punctuation. Do you have a use case in mind ? 

Anyway, if you don't want this automatic extenstion addition, you can follow the doc of `?render()`

> To avoid the automatic file extension, put the output_file value in I(), e.g., I('my-output').

<details><summary>Try to add `I()` around the file with non standard extension. </summary>

````r
dir.create(temp_proj <- tempfile())
old <- setwd(temp_proj)
ll = "---\ntitle: \"Test\"\noutput: html_notebook\n---\n\n```{r}\nplot(cars)\n```\n"
xfun::write_utf8(ll, f <- "test.Rmd")
# Case 1: Appended
rmarkdown::render(input=f, output_file="A", output_dir=".", quiet=T)
# Case 2: Not appended
rmarkdown::render(input=f, output_file="A.b", output_dir=".", quiet=T)
# Case 3: Non standard extension, so deactivating auto extension
rmarkdown::render(input=f, output_file=I("A.b_c"), output_dir=".", quiet=T)
# Long but alphanum extension is ok
rmarkdown::render(input=f, output_file="A.verylongconformingextension", output_dir=".", quiet=T)
# But with non alpha numeric character it will rename so use I()
rmarkdown::render(input=f, output_file=I("A.verylongnon-conformingextension"), output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file=I("Another.verylongnon$conformingextension"), output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file=I("A.v$x"), output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file=I("A.v=x"), output_dir=".", quiet=T)
list.files()
setwd(old)
unlink(temp_proj)
```

</details>

> # These next two understandably fail
rmarkdown::render(input=f, output_file="Another.verylongnon$conformingextension", output_dir=".", quiet=T)
rmarkdown::render(input=f, output_file="A.v$x", output_dir=".", quiet=T)
# But this one does not
rmarkdown::render(input=f, output_file="A.v=x", output_dir=".", quiet=T)

About this, this is OS specific I think. You are on linux redhad and it seems to be more constraint that Windows regarding the file extension you can use. I am on Windows and nothing fails

To sum up, I believe there is not much to do in Rmarkdown and it functionning correctly; with a mechanism offered for deactivating the auto extension. What improvment do you suggest here ? What issue did you encountered exactly ?

**Comment by rifius** [comment]
Hi Christophe,

Thanks for your prompt and detailed answer.  Mea culpa, should have refreshed my mind with the documentation before opening the ticket, though I think the restriction in `tools::file_ext()` to purely alphanumeric characters is somewhat arbitrary, when the underlying OS does not make the distinction.

Anyway, I came to report it precisely because I had a use case scenario in which I faced the behaviour: I needed to produce a number of similar reports from the same notebook source file, in each run the parameters were different and the reports would be distinguished by name.  Runs and output report names are programmatically generated by sufixing a token to the original `.Rmd` file.  I relied on the automatic adding of the extension and noticed it being added in some cases and not in others.

File naming can be quite personal, idiosyncratic and ruled by implicit convention and backwards compatibility, among other things. Now that I know what determines the behaviour I'm happy to accommodate to that idiosyncrasy.  There are plenty of examples where special characters are part of conventional file extensions

The case with `$` is a bit weird, since `for(k in files) file.create(k)` does not fail and produces the expected result in Linux.  When rendering the tests, the error message was during rename of the output file from `pandoc`: anything after the $ was missing from the name.  I would guess `pandoc` is invoked through a `system()` or `system2()` and the underlying shell is doing variable substitution on an unescaped command line.  I guess `rmarkdown` would have problems in Linux with a $ anywhere in the file name, and perhaps in Windows with a % anywhere, but I haven't touched a Windows machine in years.

To sum up, I'll close with this comment and the admonition of not having read the docs to the deepest level.  Happy with that and the `I()` cludge. If something can be suggested it'd be to lift the assumed restrictions on what is a file extension, purely on purist grounds.

Thanks,
--c
